### PR TITLE
Simplify implementation of GROUPING

### LINKED ISSUE #8109: Optimize GROUPING by re-writing it to element_at
As @martint correctly pointed out the `grouping` function is essentially doing a lookup based on the `groupId`. Since the `grouping` arguments and `grouping sets` are known at planning time, the outputs can be computed then, the function can be re-written to `element_at` where the first argument is an array of integers of the pre-computed results and the second argument is the `groupId`.

### PR DESCRIPTION
**Pull request by martint**
Instead of computing the group dynamically via a function call, pre-compute the mappings during planning.

Supersedes https://github.com/prestodb/presto/pull/8474
Fixes #8109

### CODE REVIEW

**Comment by petroav** [comment]
@martint @findepi @sopel39 I don't mind pushing this work through code review. I don't have too many things going on right now.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 1]
```diff
@@ -214,8 +214,6 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
         @Override
         public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> context)
         {
-            checkState(!SymbolsExtractor.extractUnique(context.get()).contains(node.getGroupIdSymbol()), "groupId symbol cannot be referenced in predicate");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: this and below change in this file should not belong to separate commit? IMO one thing is to inline grouping function in analyzer and the other thing is to add support for groupid be present in predicate. What do you think?

Also I think this checkState could be removed separately, as groupId cannot be used in common grouping symbols so it will be used only in wrapping filter node.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java [hunk 2]
```diff
@@ -862,6 +860,14 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Expression
 
             // Sort non-equality predicates by those that can be pushed down and those that cannot
             for (Expression conjunct : EqualityInference.nonInferrableConjuncts(inheritedPredicate)) {
+                if (node.getGroupIdSymbol().isPresent() && SymbolsExtractor.extractUnique(conjunct).contains(node.getGroupIdSymbol().get())) {
+                    // aggregation operator synthesizes outputs for group ids corresponding to the global grouping set (i.e., ()), so we
+                    // need to preserve any predicates that evaluate the group id to run after the aggregation
+                    // TODO: we should be able to infer if conditions on grouping() correspond to global grouping sets to determine whether
+                    // we need to do this for each specific case
+                    postAggregationConjuncts.add(conjunct);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: shouldn't be this block finished with `continue`, as this conjunct could be added in the code below to `pushdownConjucts` anyway?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java (hunk 1 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: Yeah, I can pull it out into a separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java]: Yes, good catch.

**Comment by martint** [comment]
@kokosing, can you take another look?
